BSP/STM32F413/Src/BSP_UART.c

```diff
@@ -172,277 +111,213 @@ static void BSP_UART3_Init() {
-            RxFifo_Peek(fifo, &data);
-        }
+uint32_t BSP_UART_Read(UART_t usart, char *str) {
+    char data = 0;
+    uint32_t recvd = 0;
 
-        RxFifo_Get(fifo, &data);
+    bool *line_recvd = lineRecvd[usart];
+    if(*line_recvd) {
+        USART_TypeDef *usart_handle = handles[usart];
 
-        *str = 0;
+        USART_ITConfig(usart_handle, USART_IT_RXNE, RESET);
+        
+        rxfifo_t *fifo = rx_fifos[usart];
+        
+        rxfifo_peek(fifo, &data);
+        while(!rxfifo_is_empty(fifo) && data != '\r') {
+            recvd += rxfifo_get(fifo, (char*)str++);
+            rxfifo_peek(fifo, &data);
+        }
 
-        fifo->lineReceived = false;
-        USART_ITConfig(uart, USART_IT_RXNE, SET);
-        return recvd;
+        rxfifo_get(fifo, &data);
+        *str = 0;
+        *line_recvd = false;
+        USART_ITConfig(usart_handle, USART_IT_RXNE, SET);
     }
 
-    return 0;
+    return recvd;
 }
 
 /**
- * @brief   Transmits data to through a specific 
- *          UART device (represented as a line of data 
- *          in csv file).
- * @param   uart device selected
- * @param   str pointer to buffer with data to send.
- * @param   len size of buffer
+ * @brief   Transmits data to through UART line
+ * @param   str : pointer to buffer with data to send.
+ * @param   len : size of buffer
+ * @param   usart : which usart to read from (2 or 3)
  * @return  number of bytes that were sent
+ * 
+ * @note This function uses a fifo to buffer the write. If that
+ *       fifo is full, this function may block while waiting for
+ *       space to open up. Do not call from timing-critical
+ *       sections of code.
  */
-uint32_t BSP_UART_Write(UART_t uartSelect, char *str, uint32_t len) {
-    USART_TypeDef *uart;
-    UartFifo_t *fifo;
-    getUSARTReferences(uartSelect, &uart, &fifo);
-
-    USART_ITConfig(uart, USART_IT_TC, RESET);
+uint32_t BSP_UART_Write(UART_t usart, char *str, uint32_t len) {
     uint32_t sent = 0;
-    while(*str != '\0' && len > 0) {
-        sent += TxFifo_Put(fifo, *str);
-        str++;
-        len--;
+
+    USART_TypeDef *usart_handle = handles[usart];
+
+    USART_ITConfig(usart_handle, USART_IT_TC, RESET);
+
+    txfifo_t *fifo = tx_fifos[usart];
+
+    while(sent < len) {
+        if(!txfifo_put(fifo, str[sent])) {
+            // Allow the interrupt to fire
+            USART_ITConfig(usart_handle, USART_IT_TC, SET);
+            // Wait for space to open up
+            while(txfifo_is_full(fifo));
+            // Disable the interrupt again
+            USART_ITConfig(usart_handle, USART_IT_TC, RESET);
+        } else {
+            sent++;  
+        }
     }
-    USART_ITConfig(uart, USART_IT_TC, SET);
+
+    USART_ITConfig(usart_handle, USART_IT_TC, SET);
+
     return sent;
 }
 
 void USART2_IRQHandler(void) {
-    UartFifo_t *fifo = &fifos[UART_2];
+#ifdef RTOS
+    CPU_SR_ALLOC();
+    CPU_CRITICAL_ENTER();
+    OSIntEnter();
+    CPU_CRITICAL_EXIT();
+#endif
```

Comment:
BPS has these ifdef guards because they write the software for bare metal as well. Controls isn't doing this, and we don't define RTOS in our build configuration, so make sure to remove these guards wherever they appear.